src/lib.rs

```diff
@@ -1,506 +1,632 @@
-    // frob the slot while we're doing the logging. This will destroy any logger
-    // set during logging.
-    let mut logger = LOCAL_LOGGER.with(|s| {
-        s.borrow_mut().take()
-    }).unwrap_or_else(|| {
-        Box::new(DefaultLogger { handle: io::stderr() }) as Box<Logger + Send>
-    });
-    logger.log(&LogRecord {
-        level: LogLevel(level),
-        args: args,
-        file: loc.file,
-        module_path: loc.module_path,
-        line: loc.line,
-    });
-    set_logger(logger);
+impl PartialOrd<LogLevel> for LogLevelFilter {
+    #[inline]
+    fn partial_cmp(&self, other: &LogLevel) -> Option<cmp::Ordering> {
+        other.partial_cmp(self).map(|x| x.reverse())
+    }
 }
 
-/// Getter for the global log level. This is a function so that it can be called
-/// safely
-#[doc(hidden)]
-#[inline(always)]
-pub fn log_level() -> u32 { unsafe { LOG_LEVEL } }
+impl Ord for LogLevelFilter {
+    #[inline]
+    fn cmp(&self, other: &LogLevelFilter) -> cmp::Ordering {
+        (*self as usize).cmp(&(*other as usize))
+    }
+}
 
-/// Replaces the task-local logger with the specified logger, returning the old
-/// logger.
-pub fn set_logger(logger: Box<Logger + Send>) -> Option<Box<Logger + Send>> {
-    let mut l = Some(logger);
-    LOCAL_LOGGER.with(|slot| {
-        mem::replace(&mut *slot.borrow_mut(), l.take())
-    })
+impl FromStr for LogLevelFilter {
+    fn from_str(level: &str) -> Option<LogLevelFilter> {
+        LOG_LEVEL_NAMES.iter()
+            .position(|&name| name.eq_ignore_ascii_case(level))
+            .map(|p| unsafe { mem::transmute(p) })
+    }
 }
 
-/// A LogRecord is created by the logging macros, and passed as the only
-/// argument to Loggers.
-#[derive(Show)]
-pub struct LogRecord<'a> {
+impl fmt::Display for LogLevelFilter {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", LOG_LEVEL_NAMES[*self as usize])
+    }
+}
 
-    /// The module path of where the LogRecord originated.
-    pub module_path: &'a str,
+impl LogLevelFilter {
+    /// Returns the most verbose logging level filter.
+    #[inline]
+    pub fn max() -> LogLevelFilter {
+        LogLevelFilter::Trace
+    }
 
-    /// The LogLevel of this record.
-    pub level: LogLevel,
+    /// Converts `self` to the equivalent `LogLevel`.
+    ///
+    /// Returns `None` if `self` is `LogLevel::Off`.
+    #[inline]
+    pub fn to_log_level(&self) -> Option<LogLevel> {
+        match *self {
+            LogLevelFilter::Off => None,
+            v => unsafe { Some(mem::transmute((v as usize) - 1)) }
+        }
+    }
+}
 
-    /// The arguments from the log line.
+/// The "payload" of a log message.
+pub struct LogRecord<'a> {
+    /// The message body.
     pub args: fmt::Arguments<'a>,
+    /// The location of the log directive.
+    pub location: &'static LogLocation,
+    /// The verbosity level of the message.
+    pub level: LogLevel,
+}
```

Comment:
In order to allow for future expansion of records in a backwards-compatible fashion, perhaps this could switch to function accessors?